`wp_get_global_stylesheet`: substitute 1-minute transient by non-persistent object cache

oandregal opened the pull request:
Trac tickets: https://core.trac.wordpress.org/ticket/56910
Related to https://github.com/WordPress/gutenberg/issues/45171

Backports the relevant parts of these PRs:

- https://github.com/WordPress/gutenberg/pull/45679
- https://github.com/WordPress/gutenberg/pull/46150

This PR:

- provides a fix to an issue reported for 6.1.1 (see trac ticket and [gutenberg issue](https://github.com/WordPress/gutenberg/issues/45713))
- backports part of the work of introducing object cache to theme.json APIs ([see](https://github.com/WordPress/gutenberg/issues/45171))

Props @mmtr @felixarntz @spacedmonkey @aristath

oandregal committed:
wp_get_global_stylesheet: substitute 1-minute transient by object cache

oandregal commented:
cc @ockham @audrasjb @costdev @hellofromtonya @ironprogrammer @mmtr @felixarntz @spacedmonkey @aristath

oandregal commented:
Sharing the context I know in case we want to test this in a plugin, first. This function is used in a couple of places:

- to enqueue the styles in the front-end ([script-loader](https://github.com/oandregal/wordpress-develop/blob/61a47155e1393035bd7c1de7864d1484d9c1ddd1/src/wp-includes/script-loader.php#L2434))
- to provide the styles to the editors ([block-editor.php](https://github.com/oandregal/wordpress-develop/blob/61a47155e1393035bd7c1de7864d1484d9c1ddd1/src/wp-includes/block-editor.php#L413))

Gutenberg substitutes the front-end styles [here](https://github.com/WordPress/gutenberg/blob/2421c97fc73a99de2bd8693db63b1c05e293df68/lib/compat/wordpress-6.1/script-loader.php#L76) and the [editor styles](https://github.com/WordPress/gutenberg/blob/2421c97fc73a99de2bd8693db63b1c05e293df68/lib/compat/wordpress-6.1/block-editor-settings.php#L64) here. I presume the test plugin should do something along those lines.

spacedmonkey reviewed (changes requested):
Same nitpitcs, but this is looking okay. Will provide profiles in time.

spacedmonkey commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -255,3 +242,42 @@ function ( $item ) {
+ * @access private
+ */
+function _wp_add_non_persistent_theme_json_cache_group() {
+	wp_cache_add_non_persistent_groups( 'theme_json' );
```
This should be here. See this example, https://github.com/WordPress/wordpress-develop/blob/acbbee8a1199f9ce70f1d3bbcc9cb80e32047e6c/src/wp-includes/load.php#L756. Add theme_json to this list.

oandregal replied:
I see `wp_cache_add_non_persistent_groups` is used in a few places:

- [load.php](https://github.com/WordPress/wordpress-develop/blob/acbbee8a1199f9ce70f1d3bbcc9cb80e32047e6c/src/wp-includes/load.php#L756)
- [class-wp-theme.php](https://github.com/oandregal/wordpress-develop/blob/61a47155e1393035bd7c1de7864d1484d9c1ddd1/src/wp-includes/class-wp-theme.php#L232)
- [ms-blogs.php](https://github.com/oandregal/wordpress-develop/blob/61a47155e1393035bd7c1de7864d1484d9c1ddd1/src/wp-includes/ms-blogs.php#L578)

I'm not sure _where_ is best. Why should we use load.php instead of having a dedicated place?

oandregal replied:
:thinking: this further. It looks like there is no a single place for this kind of thing. If so, I'd rather try to collocate `wp_cache_add_non_persistent_groups( 'theme_json' )` with the actual code that uses the cache.

The rationale for doing so is that if we move this to a different part of the codebase it becomes "out of sight, out of mind": we need to resort to human memory to remember, which is far from ideal.

oandregal replied:
Also, for cross-reference and extra clarity. I'm not sure where this should be hooked. Using `plugins_loaded` was the way we came up with at https://github.com/WordPress/gutenberg/pull/46150#discussion_r1034624190 Is there a better place?

anton-vlasenko replied:
Hmm, I wonder if it's better to move
```
add_action( 'plugins_loaded', '_wp_add_non_persistent_theme_json_cache_group' );
```
to `src/wp-includes/default-filters.php`.
System-wide hooks/actions are supposed to go into that file.
At the same time, `_wp_add_non_persistent_theme_json_cache_group()` can stay in `global-styles-and-settings.php`.

oandregal replied:
Moved the hook to `default-filters.php` in https://github.com/WordPress/wordpress-develop/pull/3712/commits/ff7fd1e3932ad7b5ec8cd8cfa66ab6054e2de165 I had missed that, thanks for bringing it up!

azaozz replied:
> I'm not sure where this should be hooked. Using plugins_loaded was the way we came up with at https://github.com/WordPress/gutenberg/pull/46150#discussion_r1034624190 Is there a better place?

Mostly depends on whether plugins may need it right away (while loading). 

The other question is: does it make sense for this to be removable by plugins? As far as I understand it, no.

As @spacedmonkey points out above the similar `wp_cache_add_non_persistent_groups( array( 'counts', 'plugins' ) );` is hard-coded and happens earlier. Thinking it makes sense for `wp_cache_add_non_persistent_groups( 'theme_json' );` to be there too. 

Then the hook and the helper function `_wp_add_non_persistent_theme_json_cache_group()` won't be needed and the only change to the code would be to add theme_json to the first one: 
```
wp_cache_add_non_persistent_groups( array( 'counts', 'plugins', 'theme_json' ) );
```
Or perhaps even better would be to add the separate `wp_cache_add_non_persistent_groups( 'theme_json' );` at the bottom of `wp_start_object_cache()` and add the nice explanation from the docblock as a comment.

oandregal replied:
> The other question is: does it make sense for this to be removable by plugins? As far as I understand it, no.

That's a good point. :thinking: 

Is there a way for it to happen earlier but still be collocated in this file?

Alternatively, if it is moved elsewhere, how can we make sure people reading this code has clarity about it being non-persistent? I can only think of adding comments in every function that uses that group.

azaozz replied:
> Is there a way for it to happen earlier but still be collocated in this file?

Don't think so unless there is a function that is called on including that file (which is generally a bad idea). Also not sure if that is particularly important. Thinking that the best option here is to add it at the bottom of `wp_start_object_cache()` which seems to be the best place for such settings, and add the explanation as inline comment. 

Many modern IDEs (probably all) would show cross references. Having a good explanation of what is being set and why, as in the current docblock, would be more than enough imho. It is easy to find. 

> can only think of adding comments in every function that uses that group

Not sure that's needed every time the cache group is used. If you're worried that somebody may miss it, perhaps adding an inline comment (pointing to where it is set) the first time it is used in global-styles-and-settings.php would make it even easier.

oandregal replied:
Done at https://github.com/WordPress/wordpress-develop/pull/3712/commits/45969f10d692f8cf83a6082e599dfcb790219118

oandregal committed:
Move hook to default-filters.php

costdev reviewed (commented):
Thanks for the PR @oandregal! Can we get some PHPUnit tests brought over from the Gutenberg PR(s)?

oandregal commented:
> Can we get some PHPUnit tests brought over from the Gutenberg PR(s)?

While https://github.com/WordPress/gutenberg/pull/45679 introduced a test for this behavior, it had to be removed later at https://github.com/WordPress/gutenberg/pull/45993 due to [late-removal of a filter](https://github.com/WordPress/gutenberg/pull/45679#discussion_r1025496394) that enabled testing this. This is an ongoing conversation happening at https://github.com/WordPress/gutenberg/issues/45912 so, I'm afraid I cannot bring any test here.

oandregal committed:
Make theme_json as non-persistent at load

spacedmonkey commented on src/wp-includes/load.php:
```diff
@@ -753,7 +753,7 @@ function wp_start_object_cache() {
 		);
 
-		wp_cache_add_non_persistent_groups( array( 'counts', 'plugins' ) );
+		wp_cache_add_non_persistent_groups( array( 'counts', 'plugins', 'theme_json' ) );
```
There are other places where `wp_cache_add_non_persistent_groups`.

https://github.com/WordPress/wordpress-develop/blob/ae42add90ef5b083368b87580fabbec1ada481f3/src/wp-includes/ms-blogs.php#L578

All places that currently have `array( 'counts', 'plugins' )`should also be updated with `, 'theme_json'`

oandregal replied:
Done at https://github.com/WordPress/wordpress-develop/pull/3712/commits/27c8ab292fab4f7a9bb00fd0a972b385028f41b5

oandregal committed:
Add test_switching_themes_recalculates_stylesheet

oandregal commented:
> Thanks for the PR @oandregal! Can we get some PHPUnit tests brought over from the Gutenberg PR(s)?

While I couldn't bring any test here for the reasons I've mentioned, I've added a new one at https://github.com/WordPress/wordpress-develop/pull/3712/commits/ef450b69ee5d0ddf33765ad5b206dd897a11f8dc. This is not perfect in that it does not test the cache, though it's still useful. When we [settle](https://github.com/WordPress/gutenberg/issues/45912) on how to do this going forward, it is easy to tweak.

oandregal committed:
Add non-persistent group to all places

oandregal commented:
@costdev @azaozz @spacedmonkey @anton-vlasenko all feedback has been addressed. What's left for this to be merged?

ironprogrammer commented:
Thanks for the updates, @oandregal! I've tested these latest PR changes from WordPress 5.9.5 and 6.0.3, and it looks good 👍🏻

In response to @oandregal:
> all feedback has been addressed. What's left for this to be merged?

While local dev testing shows this to fix the stale cache issue with inline CSS, we should wait for reporters to install the [hotfix plugin](https://github.com/ironprogrammer/wp-hotfix-56970) (or [alternate test plugin](https://github.com/ironprogrammer/wp-test-56970)) to determine if either resolves the issue in non-isolated environments.

Once more feedback has been received, it will inform us of next steps. In particular, it is important to understand how this plays with plugin or host-level caching.

costdev commented on tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:
```diff
@@ -270,4 +270,12 @@ public function test_should_enqueue_stored_styles() {
 			'Registered styles with handle of "wp-style-engine-my-styles" do not match expected value from the Style Engine store.'
 		);
 	}
+
+	public function test_switching_themes_recalculates_stylesheet() {
```
```suggestion
	/**
	 * Tests that switching themes recalculates the stylesheet.
	 *
	 * @ticket 56970
	 */
	public function test_switching_themes_should_recalculate_stylesheet() {
```

oandregal committed:
Update tests/phpunit/tests/theme/wpGetGlobalStylesheet.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

mcsf commented on tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:
```diff
@@ -270,4 +270,16 @@ public function test_should_enqueue_stored_styles() {
+	 */
+	public function test_switching_themes_should_recalculate_stylesheet() {
+		$stylesheet_for_default_theme = wp_get_global_stylesheet();
+		switch_theme( 'block-theme' );
```
Shouldn't we switch back to the default at the end of the test?

costdev replied:
Nicely spotted @mcsf!

oandregal replied:
Nice. Done at https://github.com/WordPress/wordpress-develop/pull/3712/commits/09ccca5646d446ad3da740675855bd24c37c53d1

mcsf reviewed (approved):
Looking good, @oandregal!

@ironprogrammer: I appreciate the [hotfix testing](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1350114597). How long do you estimate we should wait for feedback?

azaozz commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -83,25 +83,33 @@ function wp_get_global_styles( $path = array(), $context = array() ) {
-	$transient_name = 'global_styles_' . get_stylesheet();
+	// Ignore cache when `WP_DEBUG` is enabled, so it doesn't interfere with the theme developers workflow.
+	$can_use_cached = empty( $types ) && ! WP_DEBUG;
+	/**
```
```suggestion
	/*
```

azaozz commented:
LGTM too, just one typo (`/**` is reserved for docblocks).

oandregal committed:
Update src/wp-includes/global-styles-and-settings.php

Co-authored-by: Andrew Ozz <743931+azaozz@users.noreply.github.com>

oandregal committed:
Switch back to default theme

costdev commented on tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:
```diff
@@ -270,4 +270,19 @@ public function test_should_enqueue_stored_styles() {
+		$this->assertStringNotContainsString( '--wp--preset--font-size--custom: 100px;', $stylesheet_for_default_theme, 'custom font size (100px) not present for default theme' );
+		$this->assertStringContainsString( '--wp--preset--font-size--custom: 100px;', $stylesheet_for_block_theme, 'custom font size (100px) is present for block theme' );
+
+		switch_theme( WP_DEFAULT_THEME );
```
This should happen before the assertions in case either of them fails

oandregal replied:
Ah, good point. I was misled by the other places in the code that does the same. Resolved at https://github.com/WordPress/wordpress-develop/pull/3712/commits/62518d145322c1d1e545398f496119f1612250fb

ironprogrammer commented:
In response to @mcsf:
> @ironprogrammer: I appreciate the [hotfix testing](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1350114597). How long do you estimate we should wait for feedback?

That's a good question, and maybe best addressed by one or more core committers. Maybe @azaozz or @spacedmonkey could share some thoughts?

My personal take: While we've seen a general vote of confidence in this from core contributors, we have yet to hear back from reporters. I think it's worth allowing another couple weeks for them to test and provide feedback, since the earliest this could ship is mid-January anyways -- that is, assuming it qualifies as a fix to changes introduced in 6.1.1, and isn't pushed to 6.2 (maybe March).

azaozz commented:
>> How long do you estimate we should wait for feedback?

> ...the earliest this could ship is mid-January anyways -- that is, assuming it qualifies as a fix to changes introduced in 6.1.1

Yes, there's some time until WP 6.1.2. This PR is a fix for https://core.trac.wordpress.org/ticket/56970 that most likely will be there. However thinking it can be committed to trunk (WP 6.2-alpha) now, no need to wait. That will also facilitate some more testing (in nightly builds, etc.).

If there's feedback that requires changes, it can be addressed in another ticket/PR.

spacedmonkey reviewed (dismissed):
Sadly, in my testing, I am seeing a large performance regression. 

### Before
<img width="1792" alt="Screenshot 2022-12-15 at 23 01 30" src="https://user-images.githubusercontent.com/237508/207985223-ffecc602-4e31-4ac0-990a-1575ef9d82cf.png">

### After 
<img width="1792" alt="Screenshot 2022-12-15 at 23 02 03" src="https://user-images.githubusercontent.com/237508/207985250-0675711a-c36d-422c-90e3-030760d22f49.png">

It is 25ms slower, this is after a merge in trunk. 

This should NOT be merged until this is resolved. 

I wonder if porting https://github.com/WordPress/gutenberg/pull/46074 and https://github.com/WordPress/gutenberg/pull/46112 over first would help this issue.

felixarntz reviewed (approved):
LGTM!

felixarntz commented:
@spacedmonkey Can you provide some more context on the performance regression? Is it just because the value was already in the object cache before?

Since it is now a non persistent cache, I'd say it's expected that it is now behaving worse than before on second load. However, a 1-minute transient is arguably a less proper solution than a non-persistent cache. And on first page load performance should still be the same.

I'm onboard with the current solution, however I also think we should continue working towards an actually working _persistent_ solution to improve performance for sites that use a persistent object cache.

oandregal committed:
Switch back before assertions

oandregal commented:
I'd like to provide more context on why I think this PR should be merged without waiting for the reports of the test plugin.

This was actually created to fix this other issue https://github.com/WordPress/gutenberg/issues/43914 (not sure if there's a trac report) with the 1-minute transient. While it may also fix https://core.trac.wordpress.org/ticket/56970 (if the test plugin confirms that's the root issue), the two things are independent in my view. What if we learn the test plugin does not fix the issue? Well, in that case, we need to continue investigating _why_ it happens. But this PR would still be necessary to fix the first issue.

That's why I'd vote to merge this as early in the cycle as possible. My goal when preparing the backport was avoid the same late-minute issues we saw on past cycles by backporting really late.

oandregal commented:
> I'm onboard with the current solution, however I also think we should continue working towards an actually working persistent solution to improve performance for sites that use a persistent object cache.

:100: agree on this. Count me in to resume (actually, get back) that work after all the [high-priority stuff](https://github.com/WordPress/gutenberg/issues/45171#issuecomment-1332435270) is shipped.

oandregal commented:
_(I've updated this data on 221219-184800, see [below](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1358028676))_

In terms of performance, this PR is:

- **12ms slower** than trunk (1.70% worse) for themes with `theme.json`
- **1.79ms slower** than trunk (0.46 worse) for themes without `theme.json`

I don't know how to calculate the statistical error of running the performance analysis using xdebug, though I presume these numbers are less than the statistical error, so they can be ignored.

I've tested this using the same mechanism I use for all performance work I've been doing, so we can actually compare the impact of each PR:

- I have loaded the "Hello World!" post as a logged-out user. Conducted the test 9 times and then calculated the median ([see data in spreadsheet](https://docs.google.com/spreadsheets/d/1nMb_dqYZNzioY_g-lfXRtspRwv8SlQLg3LaUqGTzs3c/edit#gid=0)).
- Compared this PR vs trunk at `1a8535075613b29e750a3b966db4e2004dfad48d` (svn 54999).
- Using TwentyTwentyThree for theme with `theme.json` and using TwentyTwenty for a theme with no `theme.json`.

I'm also sharing the raw data I extracted (cachegrind files) so anyone can open them with their tool of choice and inspect the results. Feel free to DM me if you are interested in setting this up for running the test yourself or anything.

- Data for TwentyTwenty:
  - [twentytwenty-pr-3712.zip](https://github.com/WordPress/wordpress-develop/files/10245617/twentytwenty-pr-3712.zip)
  - [twentytwenty-trunk-1a8535075613.zip](https://github.com/WordPress/wordpress-develop/files/10245619/twentytwenty-trunk-1a8535075613.zip)
- Data for TwentyTwentyThree:
  - [twentytwentythree-trunk-1a8535075613.zip](https://github.com/WordPress/wordpress-develop/files/10245580/twentytwentythree-trunk-1a8535075613.zip)
  - [twentytwentythree-pr-3712.zip](https://github.com/WordPress/wordpress-develop/files/10245576/twentytwentythree-pr-3712.zip)

spacedmonkey commented:
I strongly believe this fix should not go into 6.1.2. This is a big change to the code and should be part of a major release. As we are effectively removing this cache by making it none persistent. We should make this part of the 6.2 release and work on other performance fixes to make the impact less on performance.

oandregal commented:
> I strongly believe this fix should not go into 6.1.2.

My understanding is that anything that goes to a patch release needs to be cherry-picked from `trunk`, and that conversation happens elsewhere (core slack channel, trac tickets for the patch release, etc.).

Is that how we do it? If my understanding is correct, what's the basis for the blocking this PR from being merged into `trunk`? Happy to learn how to unblock this.

ironprogrammer commented:
Sharing some initial reporter feedback on the hotfix 🎉: https://github.com/WordPress/gutenberg/issues/45713#issuecomment-1354333397 (and [their initial report](https://github.com/WordPress/gutenberg/issues/45713#issuecomment-1333482461)).

azaozz commented:
> This was actually created to fix this other issue with the 1-minute transient.

Generally having a 1 minute transient is not a good idea for sites without persistent cache. There transients are saved to the options table in the DB. Writing to the DB every minute from the front-end is generally considered a big bottleneck that slows down performance significantly. It also can exceed database usage limits which may result in the DB being locked down causing the site to stop working.

This is more of an edge case as sites without persistent cache usually don't get many visitors. The 1 minute DB writes may not be noticeable, or the slowness may be attributed to other factors. Similar problems have been reported before after the WP_Heartbeat API was introduced.

> I strongly believe this fix should not go into 6.1.2.

Yea, I feel "uneasy" about it too, but fixing the front-end DB writes may also be needed if there are reports by the time 6.1.2 is finalized.

> anything that goes to a patch release needs to be cherry-picked from trunk, and that conversation happens elsewhere (core slack channel, trac tickets for the patch release, etc.).

Yep, that's pretty much the workflow except the initial decision of what should be in the dot release happens earlier on Trac where the tickets are moved to the appropriate milestone. Of course all of these tickets are later reviewed, and patches are double and triple checked.

In any case I don't see anything that prevents this PR form being adding to trunk regardless of whether it will be in 6.1.2 or not.

spacedmonkey commented:
I think if this goes into core, we also need to tackle some other performance issues. If we can offset the performance regression here with some other small tweaks, then I am good to commit. What I do not want for for a 6.1.2 to make performance worse. Point releases should not make performance worse, as less people read the dev notes and will not know what is happening. 

I have highlighted these, 

> I wonder if porting https://github.com/WordPress/gutenberg/pull/46074 and https://github.com/WordPress/gutenberg/pull/46112 over first would help this issue.


But there are other tweaks like, https://github.com/WordPress/gutenberg/pull/46043

oandregal commented:
I've been peer-reviewing my own data analysis and realized I had missed setting `WP_DEBUG` to false. I fixed that and re-run the analysis. The updated data:

- themes with `theme.json` are 12ms slower (-1.70%)
- themes without `theme.json` are 1.79ms slower (-0.46%)

My conclusion is the same it was when this was a +5ms change. It's debatable this has any impact, it seems lower than the statistical error. I've added in the [data sheet](https://docs.google.com/spreadsheets/d/1nMb_dqYZNzioY_g-lfXRtspRwv8SlQLg3LaUqGTzs3c/edit#gid=688666716) a set of fields that shows the variance between the faster and the lower test: it's up to 75ms in some cases. Updated [the results above as well](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1354665003). 

Also, if we still wanted _something_ to offset this, https://github.com/WordPress/wordpress-develop/pull/3789 brings a 40% improvement for themes with `theme.json` (-350ms) and 10% improvement for themes without `theme.json` (-50ms).

oandregal commented:
I've been a bit frustrated with the variability of the data extracted from xdebug profiler. Looking at profiles I had locally from many experiments (this PR and some others), I've seen variances up to 150ms using the same steps and context. Running 10 times the same experiment reduces this a bit, but still has so much variance built in that we need to be cautious with the conclusions we take out of them.

So. Instead of looking at the cost of the full request, I'd thought I'd look at the cost of the `wp_get_global_stylesheet` before/after, given all changes here are localized to it. The cost of this function is 0.03ms in trunk and 0.04ms in this PR (both for themes with and without `theme.json`, see [data](https://docs.google.com/spreadsheets/d/1nMb_dqYZNzioY_g-lfXRtspRwv8SlQLg3LaUqGTzs3c/edit#gid=14848530)). It confirms this does not have any impact on performance.

oandregal commented:
@spacedmonkey the screenshots [you shared](https://github.com/WordPress/wordpress-develop/pull/3712#pullrequestreview-1220076911) refer to the `wp_get_global_settings` method, which the changes in this PR do not affect in any way. It sounds like it may be just part of the variability of the xdebug profiles I mentioned above?

felixarntz commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -83,25 +83,33 @@ function wp_get_global_styles( $path = array(), $context = array() ) {
-	);
-	$transient_name = 'global_styles_' . get_stylesheet();
+	// Ignore cache when `WP_DEBUG` is enabled, so it doesn't interfere with the theme developers workflow.
+	$can_use_cached = empty( $types ) && ! WP_DEBUG;
```
Similar feedback as in https://github.com/WordPress/wordpress-develop/pull/3556#discussion_r1071510762: Let's either come up with a more flexible approach on enabling this caching or remove the `! WP_DEBUG` condition entirely for now.

azaozz replied:
> or remove the `! WP_DEBUG` condition entirely

How would this code support turning off the caching when developing? Or should `SCRIPT_DEBUG` be used instead? 

> Let's either come up with a more flexible approach

That would be a "new way" to disable caching, right? If yes, it should go in a new trac ticket (enhancement). Then the above use of WP_DEBUG or SCRIPT_DEBUG can be replaced by the new method.

felixarntz replied:
Agreed, let's work on the more flexible approach in a new Trac ticket.

For now, we should add the same comment from https://core.trac.wordpress.org/browser/trunk/src/wp-includes/global-styles-and-settings.php?rev=55086#L297 above here as well. cc @hellofromtonya

oandregal replied:
Added comment at https://github.com/WordPress/wordpress-develop/pull/3712/commits/af0cb51515303ee24c90c74cbe6889cc513821ed

felixarntz commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -83,25 +83,33 @@ function wp_get_global_styles( $path = array(), $context = array() ) {
+	 * See https://github.com/WordPress/gutenberg/pull/45372
+	 */
+	$cache_group    = 'theme_json';
+	$cache_key      = 'wp_get_global_stylesheet';
```
Eventually it would be nice to also have different cache keys depending on `$types` (i.e. when it's not empty), but this can be a separate follow up. How common is it to call this function with `$types` being non-empty?

oandregal replied:
In the front-end, we only call it with no parameters ([code](https://github.com/WordPress/wordpress-develop/blob/309553e87f9b97645ec4e13cfe455af34b24abb5/src/wp-includes/script-loader.php#L2434)).

The other cases are for the editor, where we need more granularity in how we enqueue the stylesheet ([here](https://github.com/WordPress/wordpress-develop/blob/309553e87f9b97645ec4e13cfe455af34b24abb5/src/wp-includes/script-loader.php#L2953), and [here](https://github.com/WordPress/wordpress-develop/blob/309553e87f9b97645ec4e13cfe455af34b24abb5/src/wp-includes/block-editor.php#L413)) as the editor canvas (content) and the editor sidebar (inspector controls) need different pieces of information and have different CSS scopes.

spacedmonkey replied:
It would be nice to have the current theme stylesheet in this cache key too, so we don't have to do the same cache invalidation.

felixarntz commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -255,3 +258,14 @@ function ( $item ) {
+ * Private function to clean the caches under the theme_json group.
+ *
+ * @since 6.1.2
+ * @access private
+ */
+function _wp_clean_theme_json_caches() {
```
See my feedback in https://github.com/WordPress/wordpress-develop/pull/3556#discussion_r1071514752, this shouldn't be marked private.
```suggestion
 */
function wp_clean_theme_json_caches() {
```

spacedmonkey replied:
Until we finalize this api, it being private might be a good thing. We can always make this public later.

felixarntz replied:
@spacedmonkey We can't change function names in core. Even if we want to start it as private, we should still avoid the underscore prefix, that's discouraged in core anyway.

I don't see a good reason to mark this as private - it's not private anyway, and we can easily enhance what this function cleans without impacting how it's used. Pretty much all other core functions for cache cleanups are public as well.

spacedmonkey replied:
I know. But we can wrap in a public facing function. 
```
function wp_clean_theme_json_caches() {
   return _wp_clean_theme_json_caches();
}
```
This is done elsewhere in core. See 
```
	function wp_get_current_user() {
		return _wp_get_current_user();
	}
```

We can make it public later but not the other way around.

felixarntz replied:
Yes, but why even use an underscore? That is not a good pattern. Using `@access private` is the appropriate way to mark something as private.

Core has some underscore-prefixed functions, but most of them are old. In my years contributing to core several more seasoned committers have advised to not introduce underscore prefixed functions. We can always remove `@access private` later, but we can't change the name. Your workaround works, but why introduce a new function just to make something public, that isn't a good approach.

spacedmonkey replied:
> Core has some underscore-prefixed functions, but most of them are old.

- _filter_block_template_part_area - 5.9.0
- _get_block_templates_paths - 5.9.0
- _get_block_template_file - 5.9.0
- _flatten_blocks - 5.9.0
- _build_block_template_result_from_file - 5.9.0
- _wp_build_title_and_description_for_single_post_type_block_template - 6.1.0
- _wp_get_iframed_editor_assets - 6.0.0
- _resolve_home_block_template - 6.0.0

There were more examples of this, but you get the point. Underscore prefix is still actively used.

felixarntz replied:
Right, that doesn't mean it's a good practice. I did say _most_ of them are old :)

Per my explanations above, I don't think we should be using underscore prefixed functions anywhere. That's my take, based on what I learned from other committers with a lot more experience than myself.

felixarntz replied:
This is now already in core, so this comment thread is obsolete once `trunk` has been merged back into this PR.

felixarntz commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -255,3 +258,14 @@ function ( $item ) {
+/**
+ * Private function to clean the caches under the theme_json group.
+ *
+ * @since 6.1.2
```
Is this for 6.1.2 or 6.2.0?

oandregal replied:
This PR should use 6.2.

https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1385013665 for context. I guess the confusion comes from the fact that this PR may fix [an issue raised in 6.1](https://core.trac.wordpress.org/ticket/56970) for which we may (or may not) want to have a 6.1.2. But, yeah, that decision will happen in a different space, so 6.2 for now.

felixarntz reviewed (changes requested):
@oandregal Overall looks good, left just a few small pieces of feedback.

felixarntz commented:
@oandregal Is the Trac ticket reference on the PR description correct? Is this really for https://core.trac.wordpress.org/ticket/56970? Not sure I understand the relationship between the PR and ticket.

azaozz commented:
Which Gutenberg PR(s) is/are covered by this?

Are there changes in this PR that need to be backported to Gutenberg?

oandregal commented:
> Is the Trac ticket reference on the PR description correct? Is this really for https://core.trac.wordpress.org/ticket/56970? Not sure I understand the relationship between the PR and ticket.

I added the second track ticket related to this work in the issue description https://core.trac.wordpress.org/ticket/56975 

56970 is still relevant because (as far as the [experiment go](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1355742924)) using the object cache instead of the 1-minute transient also fixes an issue with styles being cache forever in some hosts.

spacedmonkey commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -255,3 +258,14 @@ function ( $item ) {
+/**
+ * Private function to clean the caches under the theme_json group.
+ *
+ * @since 6.1.2
```
```suggestion
 * @since 6.2.0
```

spacedmonkey commented:
> https://core.trac.wordpress.org/ticket/56970

Feel like https://core.trac.wordpress.org/ticket/56910 is better.

felixarntz reviewed (commented):
@oandregal @hellofromtonya Now that #3556 has been committed (via https://core.trac.wordpress.org/changeset/55086), this needs to be refreshed against `trunk`, which will make a lot of the changes in here irrelevant as they are already in core.

In regards to what this PR actually adds then, my only remaining feedback would be https://github.com/WordPress/wordpress-develop/pull/3712/files#r1073527625. After that this is good for commit as well IMO.

Let's also update the Trac ticket reference on this PR to https://core.trac.wordpress.org/ticket/56910, as suggested in https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1385274839 above.

spacedmonkey commented:
@felixarntz @oandregal  I have created a PR for making SVG function use this same caching. https://github.com/oandregal/wordpress-develop/pull/1

I believe we need to merge this as part of this change to close out. #[56910](https://core.trac.wordpress.org/ticket/56910)

hellofromtonya commented:
As @felixarntz pointed out, this PR needs a rebase and potentially a rebuild of the individual commits to remove what is no longer relevant. 

That said, there's an open discussion still happening on the caching solution (due to fatal errors in the admin area) (See the Trac ticket https://core.trac.wordpress.org/ticket/56975). Until a solution is committed, might be best to wait on the rebase.

oandregal committed:
Add comment about use of WP_DEBUG

oandregal commented:
Rebased and removed the no longer relevant parts. Changes for `src/wp-includes/load.php`, `src/wp-includes/ms-blogs.php`, `tests/phpunit/includes/abstract-testcase.php` are still relevant here as they were reverted from core at https://github.com/WordPress/wordpress-develop/pull/3864

I was unable to reproduce the bug raised at https://core.trac.wordpress.org/ticket/56975#comment:32 in this PR. I'll need to go back there and see if I can have some testing steps. I want to understand how to reproduce to either introduce the fix in this PR directly or that we don't block this from landnig if it doesn't.

As I shared in other places, this week I'm unable to contribute for more than half an hour here and there. So, go ahead when you feel is ready and don't wait for my feedback. Or ping me via DM if there's something I must absolutely review.

oandregal commented:
> @felixarntz @oandregal I have created a PR for making SVG function use this same caching. https://github.com/oandregal/wordpress-develop/pull/1

That's a good next step, thanks for working on it. As it is now, it is based on this PR. I'd recommend preparing a separate PR instead as they are unrelated changes (though, sensibly, going in the same direction). Ideally, that work happens in Gutenberg first, so it is tested by users before merging it into core.

hellofromtonya commented:
Noting: Similar work is being done in `wp_get_global_settings()` in [Trac 57502](https://core.trac.wordpress.org/ticket/57502) and https://github.com/WordPress/wordpress-develop/pull/3789. That scope of work captures the latest conversations and consensus. I'm thinking it might be worth landing PR #3789 first, then rebasing and updating this PR to remove the duplicate code, and then copy the code patterns from it.

@felixarntz @spacedmonkey @oandregal Do you agree?

oandregal commented:
I see two blocking reviews, though I think I've addressed all the feedback. Please, let me know if I should look at anything to unblock this.

felixarntz commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -83,18 +83,34 @@ function wp_get_global_styles( $path = array(), $context = array() ) {
+	$cache_key      = 'wp_get_global_stylesheet';
 	if ( $can_use_cached ) {
-		$cached = get_transient( $transient_name );
+		$cached = wp_cache_get( $cache_key, $cache_group );
```
I'm concerned that we may be reintroducing the bug from https://core.trac.wordpress.org/ticket/56975#comment:32 here. Have you tested that this problem isn't caused by adding calls to `wp_cache_*()` methods here? We need to make sure the function isn't called anywhere when `load-styles.php` is used.

hellofromtonya replied:
Thanks @felixarntz for the reminder to test this PR for the same issue that occurred with `load-styles.php` on the other PR.

I did test it. [See my test report here.](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1405464211) No fatal errors and `load-styles.php` loaded properly ✅ 

And to make sure, I added `wp_cache_get()` back into `wp_theme_has_theme_json()` and reran the tests. Tests failed as the admin area broke.

So in my testing, this PR does not affect `load-styles.php`.

Would be good to have another person test ran the same tests just to make sure. cc @ironprogrammer

felixarntz reviewed (commented):
@oandregal Overall this looks _almost_ good to go for me, however I have one potentially critical concern based on what we saw with the related #3556 approach.

hellofromtonya committed:
Updating formatting and inline comments

hellofromtonya commented:
# Test Report

As @felixarntz notes https://github.com/WordPress/wordpress-develop/pull/3712#discussion_r1088143732, need to test the PR in the WP Admin area to determine if a fatal error happens from not having `wp_cache_*()` functions available when `load-styles.php` runs.

## Testing Instructions
* Step 1: Added the following constants to `wp-config.php`:

```php
define( 'CONCATENATE_SCRIPTS', true );
define( 'WP_DEBUG', false );
define( 'SCRIPT_DEBUG', false );
```

* Step 2: Pull or apply this PR 
* Step 3: Start your localhost environment.
If using Docker, then do:
```
npm install
npm run build:dev
npm run env:start
npm run env:install
```
* Step 4: Log in and open the WP Admin area.
Does a fatal error occur?
Is the `load-styles.css` stylesheet loading?

If no, navigate through the screens. Does a fatal error occur?
If no, then this PR is not impacting `load-styles.php` ✅ 

## Testing Results

Env:
* Localhost: Docker with npm
* OS: macOS
* Browser: Chrome, Firefox, and Edge
* WP Version: `trunk`
* Plugins: none
* Theme: TT3

No fatal error ✅ 

Opened the Network tab in Dev Tools: yes, `load-styles.php` is being requested and has a 200 status ✅ 
![Screen Shot 2023-01-26 at 12 59 31 PM](https://user-images.githubusercontent.com/7284611/214925556-7d6eaa5e-3fc5-48aa-9846-3ebd508f2f25.png)

Works as expected ✅

hellofromtonya commented:
Does this cache need to be reset in `wp_clean_theme_json_cache()`?

hellofromtonya reviewed (approved):
The changes here match the previously agreed to caching strategy in changeset https://core.trac.wordpress.org/changeset/55086.

From my testing, it does not appear to impact `load-styles.php` when in the admin area.

From a code point of view, LGTM 👍  Will rely on @felixarntz and @spacedmonkey to confirm performance gains within a single request (since it's non-persistent).

Well done everyone 👏

felixarntz reviewed (approved):
Thanks @hellofromtonya for testing and verifying this is not an issue here. In that case, good to go from my end!

felixarntz commented:
@hellofromtonya 
> Does this cache need to be reset in `wp_clean_theme_json_cache()`?

Good catch, I missed that 🙃 

Can you please add the matching `wp_cache_delete` call to that function?

hellofromtonya committed:
Delete cache in wp_clean_theme_json_cache()

hellofromtonya commented:
>Can you please add the matching wp_cache_delete call to that function?

@felixarntz done in https://github.com/WordPress/wordpress-develop/pull/3712/commits/d9c0909951b2a894d137324c1082a50167a2d114 ✅ 

Retested in WP Admin. No issues 💹

ironprogrammer commented:
## Test Report
Tested with `wp-config.php` adjustments and steps outlined in https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1405464211 above.

Patch tested: this PR since https://github.com/WordPress/wordpress-develop/commit/d9c0909951b2a894d137324c1082a50167a2d114.

### Environment
- Hardware: MacBook Pro Apple M1 Pro
- OS: macOS 12.6.2
- Browser: Safari 16.2, Mozilla Firefox 109.0
- Server: nginx/1.23.3
- PHP: 7.4.33
- WordPress: 6.2-alpha-54642-src
- Theme: twentytwentythree v1.0
- Gutenberg: tested both enabled/disabled

### Actual Results
- ✅ No fatal errors occurred while navigating through or refreshing WP admin.
- ✅ Requests to `load-styles.php` returned successfully, and CSS loaded in WP admin.

hellofromtonya commented:
Thanks @ironprogrammer for testing it! 

@felixarntz should be okay to commit.

Thank you everyone for your contributions!

Commit:
Merge branch 'trunk' into update/transient-to-object-cache-in-wp-get-global-stylesheet

peterwilsoncc reviewed (commented):
Sorry to come in late. Source looks good, I just have a suggestion for an additional assertion.

peterwilsoncc commented on tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:
```diff
@@ -270,4 +270,18 @@ public function test_should_enqueue_stored_styles() {
+	 */
+	public function test_switching_themes_should_recalculate_stylesheet() {
+		$stylesheet_for_default_theme = wp_get_global_stylesheet();
+		switch_theme( 'block-theme' );
+		$stylesheet_for_block_theme = wp_get_global_stylesheet();
+		switch_theme( WP_DEFAULT_THEME );
+
```
```suggestion
		switch_theme( 'block-theme' );
		$post_switch_cache          = wp_cache_get( 'wp_get_global_stylesheet', 'theme_json' );
		$stylesheet_for_block_theme = wp_get_global_stylesheet();
		switch_theme( WP_DEFAULT_THEME );

		// This needs to be tested after switching back to the default screen, thus the variable.
		$this->assertFalse( $post_switch_cache );
```

felixarntz commented:
@peterwilsoncc Apologies, this feedback came a bit too late, I had already committed this in https://core.trac.wordpress.org/changeset/55148.

If it's crucial, we can do a follow up commit?

peterwilsoncc commented:
@felixarntz I think it should be fine without. I noticed my terrible timing ;)

azaozz commented:
Just FYI: ideally testing should happen when WP runs from `/build` as it is closest to "production" (concatenated and minified js and css, etc.). Docker/wp-env can be set to run from there. The testing instructions are slightly different:

1. `WP_DEBUG` and `SCRIPT_DEBUG` should not be set (or commented out) in wp-config.php.
2. In the terminal:
   - `npm run build` or just `grunt`. No need to do `npm install` as Grunt will run it if needed.
   - `LOCAL_DIR=build npm run env:start` to start from /build.
   - `npm run env:install` only if this is the first time using wp-env here or it will reset/empty your DB. Usually better to do `npm run env:pull` (before starting it) to update.
3. Don't forget to do `npm run env:stop` at the end.

oandregal commented:
I was able to reproduce with the [steps Tonya provided](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1405464211).

@azaozz I tried to use [the steps](https://github.com/WordPress/wordpress-develop/pull/3712#issuecomment-1405844661) you suggested, but `npm run build` doesn't work for me:

> Running "verify:old-files" task
> Warning: build/wp-includes/blocks/heading/editor.css should not be present in the $_old_files array. Use --force to continue.
> 
> Aborted due to warnings.

I tried a few things, including clearing the build folder or passing `--force` to the build command. None worked.

---


It seems this PR has been committed, so I guess it can be closed/deleted.

felixarntz commented:
Thanks all for the additional feedback. No actionable change in terms of committing per the above, so I'll indeed close this for now.

azaozz commented:
> tried to use the steps you suggested, but `npm run build` doesn't work

Uhh, sorry, this is a bug in core. Generally when you build to `/src`, it leaves the source in a "dirty" state. Trying to actually build to `/build` copies some of these left-over files and the build fails. 

There's a ticket for that already, will try to fix it soon.
